### PR TITLE
KAFKA-10813: InvalidProducerEpoch should be caught and throw as TaskMigrated

### PR DESCRIPTION
**Pull request by abbccdda**
We should catch InvalidProducerEpoch and rethrow as TaskMigrated, similar to ProducerFenced.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java [hunk 1]
```diff
@@ -184,12 +184,12 @@ public void resetProducer() {
...
     }
 
-    private void maybeBeginTransaction() throws ProducerFencedException {
+    private void maybeBeginTransaction() {
         if (eosEnabled() && !transactionInFlight) {
             try {
                 producer.beginTransaction();
                 transactionInFlight = true;
-            } catch (final ProducerFencedException error) {
+            } catch (final ProducerFencedException | InvalidProducerEpochException error) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:192]:
```suggestion
            } catch (final ProducerFencedException || InvalidProducerEpochException error) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 2]
```diff
@@ -199,7 +200,9 @@ private void recordSendError(final String topic, final Exception exception, fina
         if (isFatalException(exception)) {
             errorMessage += "\nWritten offsets would not be recorded and no more records would be sent since this is a fatal error.";
             sendException.set(new StreamsException(errorMessage, exception));
-        } else if (exception instanceof ProducerFencedException || exception instanceof OutOfOrderSequenceException) {
+        } else if (exception instanceof ProducerFencedException ||
+                exception instanceof InvalidProducerEpochException ||
+                exception instanceof OutOfOrderSequenceException) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:205]: Why is that here, we catch these three exceptions -- ProducerFenced, InvalidProducerEpoch, and OutOfOrderSequence -- and wrap them as TaskMigratedException, while in `StreamsProducer#send`, we catch  ProducerFenced, InvalidProducerEpoch, and UnknownProducerId exceptions and wrap those as TaskMigrated?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:205]: Just to clarify, I think my question/confusion is twofold:

1) Why is it OutOfOrderSequence in one place and UnknownProducerId in another?
2) What is the difference between these two code paths? Is it really possible for example for the ProducerFencedException to sometimes be thrown directly from `Producer#send`, and sometimes be passed along through the callback ( as in `streamsProducer.send(serializedRecord, (metadata, exception)` )

**Review by ableegoldman — COMMENTED** [review]
Thanks for catching this. I had a few questions since I'm not that familiar with the EOS code and exception handling I guess. Would be good if @mjsax or @guozhangwang could take a look as well

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java (hunk 2 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:205]: Wasn't over thinking here since this is a blocker-fix, so we don't want to trigger any regression here by changing existing exception catching logic. Would be good to do this as a follow-up I guess.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java (hunk 1 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:192]: This won't work I suppose?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:192]: Oh lol sorry I was just comparing it with the other block. Obviously nvm

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:205]: Ok SG, agree we should keep things simple since this is a last minute blocker

**Review by ableegoldman — APPROVED** [review]
LGTM, ok to look into the question about the different exceptions as followup work, or as part of your work to clean up exception handling in the Producer. Thanks for the fix!

**Review by mjsax — APPROVED** [review]
LGTM

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:205]: I thought about this a bit, and I think both `UnknownProducerId` and `OutOfOrderSequence` could be possibly thrown from the caller or directly (though in the later case they would be wrapped as KafkaException).

I created two tickets, one for producer and one for streams to improve the general picture moving forward.

https://issues.apache.org/jira/browse/KAFKA-10829
https://issues.apache.org/jira/browse/KAFKA-10830

**Review by guozhangwang — APPROVED** [review]

**Comment by abbccdda** [comment]
Cherry-picked to 2.7
